test/unit/BPool.t.sol

```diff
@@ -0,0 +1,557 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.23;
+
+import {BConst} from 'contracts/BConst.sol';
+import {BPool} from 'contracts/BPool.sol';
+import {IERC20} from 'contracts/BToken.sol';
+import {Test, console} from 'forge-std/Test.sol';
+import {LibString} from 'solmate/utils/LibString.sol';
+
+abstract contract Base is Test, BConst {
+  using LibString for *;
+
+  BPool public bPool;
+  address[8] public tokens;
+
+  function setUp() public {
+    bPool = new BPool();
+
+    // Create fake tokens
+    for (uint256 i = 0; i < tokens.length; i++) {
+      tokens[i] = makeAddr(i.toString());
+    }
+  }
+
+  function _preparePool(
+    BPool _pool,
+    address[8] memory _tokens,
+    uint256[8] memory _balance,
+    uint256[8] memory _denorm
+  ) internal {
+    // Create mocks
+    for (uint256 i = 0; i < _tokens.length; i++) {
+      vm.mockCall(_tokens[i], abi.encodeWithSelector(IERC20(_tokens[i]).transfer.selector), abi.encode(true));
+      vm.mockCall(_tokens[i], abi.encodeWithSelector(IERC20(_tokens[i]).transferFrom.selector), abi.encode(true));
+    }
+
+    // Set tokens
+    for (uint256 i = 0; i < _tokens.length; i++) {
+      vm.prank(_pool.getController());
+      _pool.bind(_tokens[i], _balance[i], _denorm[i]);
+    }
+
+    // Set public swap
+    vm.store(
+      address(_pool),
+      bytes32(uint256(6)),
+      bytes32(uint256(0x0000000000000000000000010000000000000000000000000000000000000000))
+    );
+    // Set finalize
+    vm.store(address(_pool), bytes32(uint256(8)), bytes32(uint256(1)));
+    // Set totalSupply
+    vm.store(address(_pool), bytes32(uint256(2)), bytes32(INIT_POOL_SUPPLY));
+  }
+
+  function _assumeHappyPath(
+    uint256[8] memory _balance,
+    uint256[8] memory _denorm
+  ) internal view returns (uint256[8] memory, uint256[8] memory) {
+    for (uint256 i = 0; i < _balance.length; i++) {
+      _balance[i] = bound(_balance[i], MIN_BALANCE, MIN_BALANCE * 1000000); // TODO: found a better max
```

Comment:
our objective is that this doesnt overflow, uint.max / x sometimes works, yet we need to find that x